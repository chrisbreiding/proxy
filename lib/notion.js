const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
