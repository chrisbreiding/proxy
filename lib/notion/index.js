const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
