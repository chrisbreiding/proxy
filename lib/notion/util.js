const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
