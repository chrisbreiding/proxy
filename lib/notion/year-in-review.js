const debug = require('debug')('proxy:scripts')
const minimist = require('minimist')

const { appendBlockChildren, getBlockChildren, getPlainText, makeBlock } = require('./util')
const { compact } = require('../util/collections')
const { patienceDiffPlus } = require('../util/patience-diff')
const { getEnv } = require('../util/env')

const notionToken = getEnv('NOTION_TOKEN')
const donePageId = getEnv('NOTION_DONE_ID')

debug('ENV:', {
  notionToken,
  donePageId,
})

const findId = (blocks, filter, error) => {
  const block = blocks.find(filter)

  if (!block && error) {
    throw new Error(error)
  }

  return block ? block.id : undefined
}

const getPageIds = async (year) => {
  const { results: doneBlocks } = await getBlockChildren({ notionToken, pageId: donePageId })

  const yearId = findId(doneBlocks, (block) => {
    return block.child_page && block.child_page.title === `${year}`
  }, `Could not find page for year: ${year}`)

  const { results: yearBlocks } = await getBlockChildren({ notionToken, pageId: yearId })

  const months = compact(yearBlocks.map((block) => {
    if (!block.child_page) return

    return {
      id: block.id,
      name: block.child_page.title,
    }
  }))

  return {
    months,
    yearId,
  }
}

const getSimilarity = (string1, string2) => {
  const diff = patienceDiffPlus(string1.split(' '), string2.split(' '))

  return (
    diff.lines.length - diff.lineCountMoved - diff.lineCountDeleted - diff.lineCountInserted
  ) / (diff.lines.length - diff.lineCountMoved)
}

const withoutEmoji = (text) => {
  return text.replace(/([\u2700-\u27BF]|[\uE000-\uF8FF]|\uD83C[\uDC00-\uDFFF]|\uD83D[\uDC00-\uDFFF]|[\u2011-\u26FF]|\uD83E[\uDD10-\uDDFF])/g, '').trim()
}

const whitespaceRegex = /\W+/g

const normalizeWhitespace = (text) => {
  return text.replace(whitespaceRegex, ' ')
}

const findMatching = (data, quest) => {
  const normalizedQuestName = normalizeWhitespace(quest)
  const questNameWithoutEmoji = withoutEmoji(normalizedQuestName)

  for (let storedQuests of data) {
    let hasSimilarity = false

    for (let storedQuest of storedQuests) {
      const normalizedStoredQuestName = normalizeWhitespace(storedQuest.quest)

      if (normalizedStoredQuestName === normalizedQuestName) {
        return { exact: true, addTo: storedQuest }
      }

      if (getSimilarity(withoutEmoji(normalizedStoredQuestName), questNameWithoutEmoji) > 0.5) {
        hasSimilarity = true
      }
    }

    if (hasSimilarity) {
      return { exact: false, addTo: storedQuests }
    }
  }
}

const dateRegex = /[A-Za-z]{3}, (\d{1,2}\/\d{1,2})/

const getMonthData = async (month, data) => {
  const { results } = await getBlockChildren({ notionToken, pageId: month.id })

  let date

  for (let block of results) {
    const text = getPlainText(block)

    if (!text) continue

    const dateMatch = text.match(dateRegex)

    if (dateMatch) {
      date = dateMatch[1]

      continue
    }

    const match = findMatching(data, text)

    if (!match) {
      data.push([{
        dates: [date],
        quest: text,
      }])

      continue
    }

    if (match.exact) {
      match.addTo.dates.push(date)
    } else {
      match.addTo.push({
        dates: [date],
        quest: text,
      })
    }
  }

  return data
}

const getDates = (quests) => quests.map((q) => q.dates).flat()

const getData = async (months) => {
  let data = []

  for (let month of months) {
    data = await getMonthData(month, data)
  }

  return data
  .filter((quests) => {
    return getDates(quests).length > 3
  })
  .map((quests) => {
    return quests.sort((a, b) => a.quest.length - b.quest.length)
  })
  .sort((a, b) => getDates(b).length - getDates(a).length)
}

const makeBlocks = (data) => {
  return data.map((quests) => {
    return makeBlock({
      text: `${quests[0].quest} (${getDates(quests).length})`,
      type: 'toggle',
      children: quests.map(({ quest, dates }) => {
        return makeBlock({
          text: `${quest} (${dates.length}) [${dates.join(', ')}]`,
          type: 'bulleted_list_item',
        })
      }),
    })
  })
}

const yearInReview = async () => {
  try {
    const args = minimist(process.argv.slice(2))

    if (!args.year) {
      throw new Error('Must specify --year')
    }

    const { months, yearId } = await getPageIds(args.year)
    const data = await getData(months)
    const blocks = makeBlocks(data)

    await appendBlockChildren({ notionToken, blocks, pageId: yearId })

    // eslint-disable-next-line no-console
    console.log('Successfully added Year In Review')

    process.exit(0)
  } catch (error) {
    // eslint-disable-next-line no-console
    console.log('Adding year failed:')
    // eslint-disable-next-line no-console
    console.log(error.stack)

    process.exit(1)
  }
}

yearInReview()
