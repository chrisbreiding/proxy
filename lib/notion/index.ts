import { addUpcomingWeek, upcomingWeekView } from './upcoming-week'
import { getBlockChildren } from './util'
import { onSocket } from './shopping'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export {
  addUpcomingWeek,
  upcomingWeekView,
  onSocket,
}
