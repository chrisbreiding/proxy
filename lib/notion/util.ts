import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
