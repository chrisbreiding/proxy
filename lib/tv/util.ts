import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
