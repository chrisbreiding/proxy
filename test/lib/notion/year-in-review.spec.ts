import { describe, expect, it } from 'vitest'

import { nockGetBlockChildren, nockAppendBlockChildren, snapshotBody } from './util'
import { yearInReview } from '../../../lib/notion/year-in-review'
import monthBlocks from '../../fixtures/notion/year-in-review/month-blocks'

describe('lib/notion/year-in-review', () => {
  it('adds year summmary', async () => {
    nockGetBlockChildren('done-page-id', { fixture: 'year-in-review/done-blocks' })
    nockGetBlockChildren('year-id', { fixture: 'year-in-review/year-blocks' })
    nockGetBlockChildren('january-id', { reply: monthBlocks.january })
    nockGetBlockChildren('february-id', { reply: monthBlocks.february })
    nockGetBlockChildren('march-id', { reply: monthBlocks.march })
    nockGetBlockChildren('april-id', { reply: monthBlocks.april })
    nockGetBlockChildren('may-id', { reply: monthBlocks.may })
    nockGetBlockChildren('june-id', { reply: monthBlocks.june })
    nockGetBlockChildren('july-id', { reply: monthBlocks.july })
    nockGetBlockChildren('august-id', { reply: monthBlocks.january })
    nockGetBlockChildren('september-id', { reply: monthBlocks.september })
    nockGetBlockChildren('october-id', { reply: monthBlocks.october })
    nockGetBlockChildren('november-id', { reply: monthBlocks.november })
    nockGetBlockChildren('december-id', { reply: monthBlocks.december })

    const snapshot = snapshotBody(nockAppendBlockChildren({ id: 'year-id' }))

    await yearInReview({
      donePageId: 'done-page-id',
      notionToken: 'notion-token',
      year: '2021',
    })

    await snapshot
  })

  it('errors if year cannot be found', async () => {
    nockGetBlockChildren('done-page-id', { fixture: 'year-in-review/done-blocks' })

    await expect(() => yearInReview({
      donePageId: 'done-page-id',
      notionToken: 'notion-token',
      year: '2025',
    })).rejects.toThrow('Could not find page for year: 2025')
  })
})
