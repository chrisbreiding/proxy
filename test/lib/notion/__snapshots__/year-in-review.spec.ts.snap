// Vitest Snapshot v1

exports[`lib/notion/year-in-review > adds year summmary 1`] = `
{
  "children": [
    {
      "object": "block",
      "toggle": {
        "children": [
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "Task 2 (14) [1/4, 3/9, 3/25, 4/5, 4/25, 5/19, 5/27, 6/15, 6/23, 1/4, 9/14, 10/4, 10/12, 11/1]",
                  "text": {
                    "content": "Task 2 (14) [1/4, 3/9, 3/25, 4/5, 4/25, 5/19, 5/27, 6/15, 6/23, 1/4, 9/14, 10/4, 10/12, 11/1]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "üçû Task 2 (17) [1/12, 1/20, 2/1, 4/13, 4/17, 6/15, 6/23, 7/5, 7/29, 1/12, 1/20, 9/6, 9/6, 9/14, 9/22, 10/20, 11/17]",
                  "text": {
                    "content": "üçû Task 2 (17) [1/12, 1/20, 2/1, 4/13, 4/17, 6/15, 6/23, 7/5, 7/29, 1/12, 1/20, 9/6, 9/6, 9/14, 9/22, 10/20, 11/17]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
        ],
        "rich_text": [
          {
            "plain_text": "Task 2 (31)",
            "text": {
              "content": "Task 2 (31)",
            },
            "type": "text",
          },
        ],
      },
      "type": "toggle",
    },
    {
      "object": "block",
      "toggle": {
        "children": [
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "Task 1 (10) [1/4, 2/9, 3/25, 4/5, 4/17, 7/5, 7/13, 7/21, 1/4, 9/14]",
                  "text": {
                    "content": "Task 1 (10) [1/4, 2/9, 3/25, 4/5, 4/17, 7/5, 7/13, 7/21, 1/4, 9/14]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "üêü Task 1 (16) [1/20, 3/17, 4/13, 4/25, 7/5, 7/29, 7/29, 1/20, 9/6, 10/20, 10/20, 10/20, 10/28, 11/9, 11/25, 12/6]",
                  "text": {
                    "content": "üêü Task 1 (16) [1/20, 3/17, 4/13, 4/25, 7/5, 7/29, 7/29, 1/20, 9/6, 10/20, 10/20, 10/20, 10/28, 11/9, 11/25, 12/6]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
        ],
        "rich_text": [
          {
            "plain_text": "Task 1 (26)",
            "text": {
              "content": "Task 1 (26)",
            },
            "type": "text",
          },
        ],
      },
      "type": "toggle",
    },
    {
      "object": "block",
      "toggle": {
        "children": [
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "Task 3 (11) [1/12, 2/9, 2/17, 3/1, 4/25, 6/7, 6/7, 6/15, 1/12, 11/1, 12/14]",
                  "text": {
                    "content": "Task 3 (11) [1/12, 2/9, 2/17, 3/1, 4/25, 6/7, 6/7, 6/15, 1/12, 11/1, 12/14]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "üå≤ Task 3 (14) [1/28, 3/9, 3/25, 4/5, 5/3, 5/11, 5/19, 5/19, 6/23, 7/29, 1/28, 9/14, 10/28, 12/14]",
                  "text": {
                    "content": "üå≤ Task 3 (14) [1/28, 3/9, 3/25, 4/5, 5/3, 5/11, 5/19, 5/19, 6/23, 7/29, 1/28, 9/14, 10/28, 12/14]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
        ],
        "rich_text": [
          {
            "plain_text": "Task 3 (25)",
            "text": {
              "content": "Task 3 (25)",
            },
            "type": "text",
          },
        ],
      },
      "type": "toggle",
    },
    {
      "object": "block",
      "toggle": {
        "children": [
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "Task 4 (4) [1/28, 5/19, 6/15, 1/28]",
                  "text": {
                    "content": "Task 4 (4) [1/28, 5/19, 6/15, 1/28]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
          {
            "bulleted_list_item": {
              "rich_text": [
                {
                  "plain_text": "üí¨ Task 4 (21) [1/12, 1/20, 2/9, 2/17, 3/1, 3/17, 5/3, 5/11, 5/11, 6/7, 6/7, 7/21, 7/21, 1/12, 1/20, 10/12, 10/28, 11/9, 11/17, 11/25, 12/22]",
                  "text": {
                    "content": "üí¨ Task 4 (21) [1/12, 1/20, 2/9, 2/17, 3/1, 3/17, 5/3, 5/11, 5/11, 6/7, 6/7, 7/21, 7/21, 1/12, 1/20, 10/12, 10/28, 11/9, 11/17, 11/25, 12/22]",
                  },
                  "type": "text",
                },
              ],
            },
            "object": "block",
            "type": "bulleted_list_item",
          },
        ],
        "rich_text": [
          {
            "plain_text": "Task 4 (25)",
            "text": {
              "content": "Task 4 (25)",
            },
            "type": "text",
          },
        ],
      },
      "type": "toggle",
    },
  ],
}
`;
